contributor-docs/adrs/adr-001-custom-elements.md

```diff
@@ -0,0 +1,115 @@
+
+### Organizational Overhead
+- GitHub’s custom elements are all managed in different repos which introduces more maintenance overhead.
+- Reacting to changes will take a bit more time, as we’ll need to wait for changes to be made to the custom elements until we can update Primer React Components (versus having behavior Hooks in PRC that we can update at our leisure).
+- Engineers who want to contribute to Primer React Components to build new components and behaviors would need to be familiar with both custom elements and React, two very different paradigms, and context switch between the two.
+
+### Other
+- The custom element & web component API progress slower than React due to changes needing to go through the whatwq standards process.
+
+### Risks of not switching to custom elements for behaviors
+- We spend extra time building behaviors in React that have already been built in our [custom elements library](https://github.github.io/web-systems-documentation/#custom-elements).
+  - There are currently 19 behaviors/components listed on the custom elements documentation site. Several of these we have already implemented in React in either PRC, Doctocat, or other React applications at GitHub which can be upstreamed (details-dialog, details-menu, clipboard-copy, text-expander, autocomplete, task-list via drag and drop hooks, tab-container, text-expander).
+- We decide not to invest further in React at GitHub and have wasted time we could have spent building more custom elements.
+  - This seems unlikely as there seems to be clear consensus that we will continue to build more and more highly interactive products.
+- The React library is abandoned and becomes obsolete
+  - This is a risk with any technology that we may use, seems highly unlikely in the near term, and is also a risk for custom elements.
+- Behaviors in .com using custom elements and behaviors in PRC diverge, leading to an inconsistent experience
+  - This is probably the biggest risk we face, but moving to custom elements isn’t necessarily the only or best solution. We should explore other ways of detecting divergence such as integration tests.
+
+### Decision
+Due to the challenges listed above and our priorities listed in the [Assumptions](#assumptions) section, we are not investing time in building out behaviors with custom elements in our Primer React Components library. Instead, we should spend time expanding coverage using React Hooks and focus on finding other approaches for making sure implementation of behaviors in our different stacks are consistent (such as integration tests).
```

Comment:
This seems to enumerate the reasons for not implementing behaviours in Web Components but fails to address the reasons why we _should_ implement behaviours using React Hooks.

Implementing behaviours in React Hooks couples them to React. Another option for us is to write functions which return predefined interfaces that can be integrated with _either_ hooks _or_ Web Components. Both Web Component's lifecycle methods and React Hooks are effectively interchangeable injection points (e.g. `useEffect` is analogous to `connectedCallback`). Why would we not explore this avenue further?

Reply:
> fails to address the reasons why we should implement behaviours using React Hooks

@T-Hugs brought up the point yesterday that we could also test out writing behaviors in vanilla JS to be reused across any UI stack at GitHub. This approach would still face the same organizational & maintenance challenges I've detailed in this ADR.

Reply:
> Another option for us is to write functions which return predefined interfaces that can be integrated with either hooks or Web Components. Both Web Component's lifecycle methods and React Hooks are effectively interchangeable injection points (e.g. useEffect is analogous to connectedCallback). Why would we not explore this avenue further?

I think this is outside the scope of this ADR. I am working on exploring this.

Reply:
@T-Hugs @keithamus  I might be still missing some context around what you two discussed last week regarding [strategy alignment](https://github.slack.com/archives/C01C60YF4BE/p1611249458000700?thread_ts=1611227360.000200&cid=C01C60YF4BE). From what I gathered it sounded like we were all in alignment with using React Hooks & React state to manage behaviors. Can one of you write up a summary of what you discussed? Behavior work is scheduled to start next week, and we have deadlines relating to big ships later this year so changing strategy now is going to change the epic we've planned quite a bit. Either way, we need to work quickly to get into alignment.